### PR TITLE
Add post regenerations

### LINKED ISSUE #4623: Give autogenerate IQDB/sample/preview capability to moderators
As seen from #3386, this is a recurrent issue. Giving moderators this capability would allow for them to share in the work and would most likely quicken the response to any issues. I believe that the image replacement function includes both of these regeneration capabilities, but that seems like an awful lot of spam will be added to the image replacement index, considering that there is currently 250K+ posts that need to be fixed.

This could also be made granular, to reduce the load on the servers, but also to accurately document the fixes that occur on each post.

* Example: 
```
Method: POST
URL: /post_regenerations
Body: 
{
    post_id: 1234,
    type: iqdb
}
```

The different types should be **iqdb**, **sample**, and **preview**. The only thing the records would need to store are the post ID, the type, the timestamps, and the creator.

### PR DESCRIPTION
**Pull request by BrokenEagle**
Fixes #4623.

The controls are presented as links in the **Options** section to moderators on the post show page.  It's also available to be used from the API as well which will help with scripting. Some of the code in the uploader service was further functionalized so that it could be reused by the post regenerator.

### CODE REVIEW

File: app/models/post_regeneration.rb [hunk 1]
```diff
@@ -0,0 +1,37 @@
+class PostRegeneration < ApplicationRecord
+  belongs_to :creator, :class_name => "User"
+  belongs_to :post
+
+  validates_inclusion_of :category, :in => %w(iqdb resizes)
```
[the remark this answers is not in the document]
Comment (BrokenEagle) [reply app/models/post_regeneration.rb]: Fixed in 98cf4c9.

[the remark this answers is not in the document]
Comment (BrokenEagle) [reply app/models/post_regeneration.rb]: Fixed in 98cf4c9.

**Comment by BrokenEagle** [comment]
Did a fix on the style issues and rebased it after the update on the master branch to `db/structure.sql` .

**Comment by BrokenEagle** [comment]
I just wanted to note that there was some weirdness with the tests when trying out the `resizes` option, which is why it wasn't included in the tests. It was trying to regenerate the images using the files of a previous run, and since those files were no longer there, it would cause an exception.

**Comment by BrokenEagle** [comment]
Fixed the merge conflict. Also updated the model file for e771c0fca820dadafe40fa72a3d4f6364da8f27e and ee4516f5fe6e665e98d64a0bdde48208a59d4afb.

**Comment by evazion** [comment]
I merged this in 16d6f3bbd5951e11578f85fe019f44edafc0f309 and made some other changes in df44937c57b12389ec9331123ef59ee7980bb576, b6f9c9a866a5cced78288d9251708ff87557da51, and 79ee6a515de79d11b1140e3b26a5bb6225e31384:

* I removed the PostRegeneration model and changed it to just log a mod action when instead. I don't really want an entire model and all the other stuff that brings (extra history page, extra index page, tests, etc) just to log a simple "user A regenerated post B" message.
* I changed it so that regenerations happen asynchronously in a delayed job. Regenerations can be slow and I don't want them to block the web server.
* I changed it so that the default is to regenerate both the images and IQDB. You have to use the API if you want to regenerate IQDB only. This is something normal users shouldn't have to do anyway.
* I changed it to purge the old images from Cloudflare after the images are regenerated.

**Pull request closed without merging**